Fix IJON fastresume functionality

vi3tL0u1s opened the pull request:
I am also working on removing unnecessary logics!

vanhauser-thc committed:
Merge pull request #2533 from AFLplusplus/dev

push to stable

vi3tL0u1s committed:
Fix IJON fastresume functionality

vanhauser-thc commented on src/afl-fuzz.c:
```diff
@@ -2641,8 +2641,11 @@ int main(int argc, char **argv_orig, char **envp) {
-    afl->fsrv.map_size -= MAP_SIZE_IJON_BYTES;
-    afl->fsrv.real_map_size -= MAP_SIZE_IJON_BYTES;
+    // Only adjust map_size if not doing fastresume (fastresume will handle map_size correctly)
+    if (!fast_resume) {
```
why not for fast-resume?

vi3tL0u1s replied:
Sorry, I missed this comment yesterday.

During a fresh session, we subtract the IJON bytes from the map size to reserve space for IJON. In fast-resume, however, the saved state already includes the adjusted map sizes from the original run. If we apply the subtraction again, we'd end up double-adjusting, which shrinks the map incorrectly. This check makes sure the adjustment only happens once, keeping the save/load behavior consistent.

vanhauser-thc commented on src/afl-fuzz.c:
```diff
@@ -2654,10 +2657,11 @@ int main(int argc, char **argv_orig, char **envp) {
     ck_free(max_dir);
 
-    setenv("AFL_NO_IJON", "1", 1);
+    // setenv("AFL_NO_IJON", "1", 1);
```
no this is essential. we do not need ijon extra map for cmplog and asan

vanhauser-thc commented on src/afl-fuzz.c:
```diff
@@ -2864,52 +2868,50 @@ int main(int argc, char **argv_orig, char **envp) {
 
-            }
+        // Enable IJON for fastresume (will be confirmed by forkserver handshake later)
+        afl->fsrv.use_ijon = 1;
```
that is after the forkserver handshake - not before

vi3tL0u1s committed:
Fix IJON AFL_NO_IJON handling and preventing double map_size adjustment

vi3tL0u1s committed:
Remove unnecessary struct fields

vanhauser-thc commented:
is this done and you have verified that it is working?
Or is there still something missing or needs testing?

vanhauser-thc commented on src/afl-fuzz-run.c:
```diff
@@ -156,18 +156,22 @@ fsrv_run_result_t __attribute__((hot)) fuzz_run_target(afl_state_t      *afl,
+        // Use saved IJON offset directly during fastresume for consistent offset calculation  
+        dynamic_shared_access_t *shared_access;
+        if (has_saved_ijon_state()) {
+          ijon_fastresume_state_t *saved_state = get_saved_ijon_state();
```
Why this section here? this is executed after every single fuzz.

vanhauser-thc committed:
Merge branch 'AFLplusplus:stable' into ijon-implementation

vanhauser-thc committed:
code format

vanhauser-thc commented:
note that I have pushed code formatting

vanhauser-thc commented:
please remove not-needed commented out code.

is this done and you have verified that it is working?
Or is there still something missing or needs testing?

vi3tL0u1s committed:
Move shared_access setup to initialization, and add ijon_shared_access field to afl_state_t structure for persistent storage

vi3tL0u1s commented:
I've cleaned up the architecture: the IJON shared_access setup was moved out of the code path in run_target() into proper one-time initialization, added ijon_shared_access to afl_state_t for persistent storage, and removed leftover debug code. This avoids repeated setup on every fuzz execution while keeping everything fully functional. Tests confirmed it works correctly with Mario+CmpLog+IJON and Maze+ASAN+IJON across multiple fastresume sessions.

vanhauser-thc commented:
OK I will merge this and then push to stable so hopefully some people test it
